Gdi+ capability check revision (clean)

gvheertum opened the pull request:
### Reference Issue

Resubmit of #511 
Following the remarks in #502 by @H1Gdev

### What does this implement/fix? Explain your changes.

Added a function to query gdi+ capabilities in the SVGDocument and changed the working of the document load to call an ensure function that will throw the gdi exception instead of having a large try-catch block with multiple exception type queries. The new functions are "tested" in 2 simple unittests that allows developers to check their capabilities by executing the tests.

Also updated the csproj of the SVG project to only include the mac version for the targeted framework (I encoutered some issues with targeting).

mrbean-bremen commented:
Thanks! I will have a closer look tonight.
@H1Gdev - please have a look, too.

mrbean-bremen commented on Source/SvgDocument.cs:
```diff
@@ -175,6 +175,68 @@ public virtual SvgElement GetElementById(string id)
+                    // Throw only the customized exception if we are sure GDI+ is causing the problem
+                    throw new SvgGdiPlusCannotBeLoadedException(e);
+                }   
+                //Ignore any other form of error
```
I would rather re-raise here, as there must be something really wrong if this raises.

gvheertum replied:
Fair point, I would not expect it to happen, but I share the vision that this should yield an error for the caller to handle.

mrbean-bremen commented on Source/SvgDocument.cs:
```diff
@@ -280,181 +342,143 @@ public static T Open<T>(Stream stream, Dictionary<string, string> entities) wher
+            var styles = new List<ISvgNode>();
 
-                while (reader.Read())
+            while (reader.Read())
```
This is now outside the exception handler. I this intentional?

gvheertum replied:
I checked the older versions, and it seems that prior my changes this code was outside of the try catch. During the initial changes I was not sure where the issues regarding GDI+ were raised, so I put the whole call chain in a try/catch, with additions to the original catch blocks to validate the exception type. Since we now perform the Ensure function at the start of the call we no longer need the try catch construction regarding the GDI+ errors. Other errors are not to be expected, so I thinks it's safe to leave it out, the errors will be raised to the caller since I am not sure how and what we should handle if we put this in a big try/catch.

mrbean-bremen commented on Tests/Svg.UnitTests/GdiPlusTests.cs:
```diff
@@ -0,0 +1,26 @@
+        public void GdiPlus_EnsureCapability_YieldsNoError()
+        {
+            SvgDocument.EnsureSystemIsGdiPlusCapable();
+            //This should not cause problems
```
Don't understand the comment... does it belong to the line above?

gvheertum replied:
Yeah, it's my silly way to say, "the above call is void and we should arrive here safely", I updated this part.

H1Gdev commented on Tests/Svg.UnitTests/GdiPlusTests.cs:
```diff
@@ -0,0 +1,26 @@
+        }
+
+    }
+}
```
Please end with an empty line.

gvheertum replied:
Updated, was not aware of that convention.

H1Gdev commented on Source/SvgDocument.cs:
```diff
@@ -175,6 +175,68 @@ public virtual SvgElement GetElementById(string id)
+        {
+            var currE = e;
+            int cnt = 0; // Keep track of depth to prevent endless-loops
+            while (currE != null && cnt < 10)
```
Have you exceeded 10 in your experience ?

gvheertum replied:
Unfortunately yes (not in this library), since we get an error from a deeper library, we cannot always be sure that the content of the exception is "safe". There is a slight possiblity that somewhere somebody wired 2 exceptions as innerexceptions of each other causing an loop, also if we encounter issues like stack of memory issues, there is a possibility that the exception has a lot of inner exceptions, so I guess it's safe to stop at 10. I won't expect it to be deeper than 4 or 5, but if we have to look beyond 10 children, I think we can assume we have a bigger issues (and most likely not a GDI+ issue).

H1Gdev commented:
@gvheertum 

I think that this PR has become very easy to understand !!
Thanks !

gvheertum commented:
@mrbean-bremen and @H1Gdev, I added some comments and updated the branch with a few commits to resolve the remarks/comments. Good to hear that the pushed changes were well received and made the GDI+ check more readable/usable.

Could you have another look and resolve the conversation (not sure if I should do that if I pushed an update). Thanks. :)
